### PR TITLE
[MIG] stock_reserve 15.0

### PR DESCRIPTION
**Pull request by flachica**

### CODE REVIEW

**Comment by victoralmau** [comment]
Thank you for your work.
There is something wrong with this migration, the `stock_reserve_sale` addon is added and it should not.

Maybe it would be a good idea to finish the migration to v14 (https://github.com/OCA/stock-logistics-warehouse/pull/1346) and then continue with this one.

**Comment by flachica** [comment]
Sorry @victoralmau, this module is migrated in #1486. I got confused with branch

**Comment by rousseldenis** [comment]
@flachica Could you remove unrelated files (stock_reserve_sale) ?

**Comment by flachica** [comment]
> @flachica Could you remove unrelated files (stock_reserve_sale) ?

@rousseldenis In this PR I got confused about the branch but I thought I had solved it. Please, can you tell me where you see the remaining files? I can not find it

**Comment by rousseldenis** [comment]
> > @flachica Could you remove unrelated files (stock_reserve_sale) ?
> 
> @rousseldenis In this PR I got confused about the branch but I thought I had solved it. Please, can you tell me where you see the remaining files? I can not find it

No, it's ok, wrong browser tab :sweat:

File: stock_reserve/model/stock_reserve.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    @api.onchange("product_uom_qty")
+    def _onchange_quantity(self):
+        """On change of product quantity avoid negative quantities"""
+        if not self.product_id or self.product_uom_qty <= 0.0:
+            self.product_uom_qty = 0.0
+
+    def open_move(self):
+        self.ensure_one()
+        action = self.env.ref("stock.stock_move_action")
+        action_dict = action.read()[0]
```
Comment (victoralmau) [inline stock_reserve/model/stock_reserve.py]:
```suggestion
        action_dict = self.env["ir.actions.act_window"]._for_xml_id(
            "stock.stock_move_action"
        )
```

File: stock_reserve/model/product.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        for product in self:
+            product.reservation_count = sum(
+                product.product_variant_ids.mapped("reservation_count")
+            )
+
+    def action_view_reservations(self):
+        self.ensure_one()
+        ref = "stock_reserve.action_stock_reservation_tree"
+        product_ids = self.mapped("product_variant_ids.id")
+        action_dict = self.env.ref(ref).read()[0]
```
Comment (victoralmau) [inline stock_reserve/model/product.py]:
```suggestion
       action_dict = self.env["ir.actions.act_window"]._for_xml_id(
            "stock_reserve.action_stock_reservation_tree"
        )
        product_ids = self.mapped("product_variant_ids.id")
```

File: stock_reserve/model/product.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                ("product_id", "=", product.id),
+                ("state", "in", ["draft", "assigned"]),
+            ]
+            reservations = self.env["stock.reservation"].search(domain)
+            product.reservation_count = sum(reservations.mapped("product_qty"))
+
+    def action_view_reservations(self):
+        self.ensure_one()
+        ref = "stock_reserve.action_stock_reservation_tree"
+        action_dict = self.env.ref(ref).read()[0]
```
Comment (victoralmau) [inline stock_reserve/model/product.py]:
```suggestion
        action_dict = self.env["ir.actions.act_window"]._for_xml_id(
            "stock_reserve.action_stock_reservation_tree"
        )
```

File: stock_reserve/tests/test_stock_reserve.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        self.product = product_form.save()
+        self.env["stock.quant"].create(
+            {
+                "product_id": self.product.id,
+                "location_id": self.warehouse.lot_stock_id.id,
+                "quantity": 10.0,
+            }
+        )
+
+    def test_reservation_and_reservation_release(self):
```
Comment (victoralmau) [inline stock_reserve/tests/test_stock_reserve.py:34]:
```suggestion
    def _create_stock_reservation(self, qty):
        form_reservation = Form(self.env["stock.reservation"])
        form_reservation.product_id = self.product
        form_reservation.product_uom_qty = qty
        form_reservation.location_id = self.warehouse.lot_stock_id
        return form_reservation.save()
    def test_reservation_and_reservation_release(self):
```

File: stock_reserve/tests/test_stock_reserve.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        form_reservation_1 = Form(self.env["stock.reservation"])
+        form_reservation_1.product_id = self.product
+        form_reservation_1.product_uom_qty = 6
+        form_reservation_1.location_id = self.warehouse.lot_stock_id
+        reservation_1 = form_reservation_1.save()
```
Comment (victoralmau) [inline stock_reserve/tests/test_stock_reserve.py]:
```suggestion
        reservation_1 = self._create_stock_reservation(6)
```

File: stock_reserve/tests/test_stock_reserve.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        reservation_1.reserve()
+        self.assertEqual(self.product.virtual_available, 4)
+        form_reservation_2 = Form(self.env["stock.reservation"])
+        form_reservation_2.product_id = self.product
+        form_reservation_2.product_uom_qty = 1
+        form_reservation_2.location_id = self.warehouse.lot_stock_id
+        reservation_2 = form_reservation_2.save()
```
Comment (victoralmau) [inline stock_reserve/tests/test_stock_reserve.py]:
```suggestion
        reservation_2 = self._create_stock_reservation(1)
```

File: stock_reserve/tests/test_stock_reserve.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        cron = self.env.ref("stock_reserve.ir_cron_release_stock_reservation")
+        cron.method_direct_trigger()
+        self.assertEqual(self.product.virtual_available, 10)
+
+    def test_cron_reserve(self):
+        form_reservation_1 = Form(self.env["stock.reservation"])
+        form_reservation_1.product_id = self.product
+        form_reservation_1.product_uom_qty = 11
+        form_reservation_1.location_id = self.warehouse.lot_stock_id
+        reservation_1 = form_reservation_1.save()
```
Comment (victoralmau) [inline stock_reserve/tests/test_stock_reserve.py]:
```suggestion
        reservation_1 = self._create_stock_reservation(11)
```

File: stock_reserve/tests/test_stock_reserve.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        reservation_1.release_reserve()
+        self.assertEqual(self.product.virtual_available, 9)
+
+    def test_cron_release(self):
+        form_reservation_1 = Form(self.env["stock.reservation"])
+        form_reservation_1.product_id = self.product
+        form_reservation_1.product_uom_qty = 6
+        form_reservation_1.location_id = self.warehouse.lot_stock_id
+        form_reservation_1.date_validity = fields.Date.from_string("2021-01-01")
+        reservation_1 = form_reservation_1.save()
```
Comment (victoralmau) [inline stock_reserve/tests/test_stock_reserve.py]:
```suggestion
        reservation_1 = self._create_stock_reservation(6)
        reservation_1.date_validity = = fields.Date.from_string("2021-01-01")
```

**Review by victoralmau — CHANGES_REQUESTED** [review]

**Comment by flachica** [comment]
@victoralmau How I can set this commit as coauthored with you? You are made a lot of valueable suggestions

**Comment by rousseldenis** [comment]
> @victoralmau How I can set this commit as coauthored with you? You are made a lot of valueable suggestions

You can use git cherry-pick to do so. Or oca-port tool https://pypi.org/project/oca-port/

**Review by victoralmau — APPROVED** [review]
Code review OK.

IMO, you don't need to add me as a co-author.

**Comment by CasVissers-360ERP** [comment]
@rousseldenis Think this one should be ready to merge?

**Review by rousseldenis — APPROVED** [review]
Code review

**Pull request merged**